parcels/kernel.py

```diff
@@ -378,10 +378,14 @@ def execute(self, pset, endtime, dt, recovery=None, output_file=None, execute_on
 
         def remove_deleted(pset):
             """Utility to remove all particles that signalled deletion"""
-            indices = pset.particle_data['state'] == OperationCode.Delete
-            if np.count_nonzero(indices) > 0 and output_file is not None:
-                output_file.write(pset, endtime, deleted_only=indices)
-            pset.remove_booleanvector(indices)
+            # Indices marked for deletion.
+            bool_indices = np.array([
+                p.state == OperationCode.Delete for p in pset])
+            indices = np.where(bool_indices)[0]
```

Comment:
these are 2 separate loopings which effect performance negatively - why change that ?!